### PR TITLE
Added a new default gauge that contains build info.

### PR DESCRIPTION
**Pull request by djloude**
Edited some source code to add a new default metric. cloudwatch_exporter_build_info contains two labels, 1. build_version, 2. release_date. This new gauge metric will have a default value of 1 but the label values correspond to new private class variables respectively named BuildVersion and ReleaseDate. These variables WILL need to updated prior to a new release.
This update is in response to https://github.com/prometheus/cloudwatch_exporter/issues/387.

### CODE REVIEW

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java [hunk 1]
```diff
@@ -699,6 +683,28 @@ private void scrape(List<MetricFamilySamples> mfs) {
       samples.add(new MetricFamilySamples.Sample(
           "cloudwatch_exporter_scrape_error", new ArrayList<>(), new ArrayList<>(), error));
       mfs.add(new MetricFamilySamples("cloudwatch_exporter_scrape_error", Type.GAUGE, "Non-zero if this scrape failed.", samples));
+
+      /**
+       * New static cloudwatch_exporter_build_info implementation. Sets a gauge to 1 and creates two label/value
+       * pairs that can be displayed (Grafana style UI).
+       * NOTE - When releasing a new version the private class variables must be set. (lines 31 and 32).
```
Comment (matthiasr) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java:690]: This is not great, I *will* forget to do that. Maven knows the version when building, is there any way that can be passed in during the build?

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java [hunk 2]
```diff
@@ -3,52 +3,34 @@
...
-import software.amazon.awssdk.services.cloudwatch.model.Datapoint;
-import software.amazon.awssdk.services.cloudwatch.model.Dimension;
-import software.amazon.awssdk.services.cloudwatch.model.DimensionFilter;
-import software.amazon.awssdk.services.cloudwatch.model.GetMetricStatisticsRequest;
-import software.amazon.awssdk.services.cloudwatch.model.GetMetricStatisticsResponse;
-import software.amazon.awssdk.services.cloudwatch.model.ListMetricsRequest;
-import software.amazon.awssdk.services.cloudwatch.model.ListMetricsResponse;
-import software.amazon.awssdk.services.cloudwatch.model.Metric;
-import software.amazon.awssdk.services.cloudwatch.model.Statistic;
+import software.amazon.awssdk.services.cloudwatch.model.*;
```
Comment (matthiasr) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java:11]: Cleaning this up makes sense, but could you submit it as a separate PR please?

**Review by matthiasr — DISMISSED** [review]

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java (hunk 1 above)
Comment (djloude) [reply src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java:690]: Work in progress, but i'm not very familiar with maven. In the  pom.xml file that is used to build the project do you manually set the new version in the <version> tag or is it done dynamically?

**Comment by matthiasr** [comment]
Currently this is updated by the maven release plugin during the release
process. It is the one to read

**Comment by brian-brazil** [comment]
https://github.com/prometheus/jmx_exporter/pull/279 may be of use.

File: README.md [hunk 3]
```diff
@@ -110,7 +110,22 @@ All metrics are exported as gauges.
 
 In addition `cloudwatch_exporter_scrape_error` will be non-zero if an error
 occurred during the scrape, and `cloudwatch_exporter_scrape_duration_seconds`
-contains the duration of that scrape.
+contains the duration of that scrape. `cloudwatch_exporter_build_info` will be
+non-zero if an error occurred scraping the build info.
+
+### Build Info Metric (New)
+`cloudwatch_exporter_build_info` is a new default cloudwatch exporter metric that contains the current
```
Comment (matthiasr) [inline README.md:117]:
```suggestion
### Build Info Metric

`cloudwatch_exporter_build_info` is a default cloudwatch exporter metric that contains the current
```

The README describes the current state, we don't have to say "new" here (and it would be out of date one release later anyway).

File: src/main/resources/.properties [hunk 4]
```diff
@@ -0,0 +1,2 @@
+ReleaseDate=${project.releaseDate}
+BuildVersion=${project.version}
```
Comment (matthiasr) [inline src/main/resources/.properties:2]:
```suggestion
BuildVersion=${project.version}
```

I (and git) like trailing newlines 😄

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java [hunk 5]
```diff
@@ -699,6 +687,34 @@ private void scrape(List<MetricFamilySamples> mfs) {
...
+
+      labelNames.add("build_version");
+      labelValues.add(buildVersion);
+      labelNames.add("release_date");
+      labelValues.add(releaseDate);
+
+      samples = new ArrayList<>();
+      samples.add(new MetricFamilySamples.Sample(
+          "cloudwatch_exporter_build_info", labelNames, labelValues, errorFlag));
+      mfs.add(new MetricFamilySamples("cloudwatch_exporter_build_info", Type.GAUGE, "Non-zero if build info scrape failed.", samples));
```
Comment (matthiasr) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java:716]: the build info metric should always have value `1`. If fetching the build info fails, make the label values `unknown` to match how other build info metrics behave.

File: src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java [hunk 6]
```diff
@@ -699,6 +687,34 @@ private void scrape(List<MetricFamilySamples> mfs) {
...
       mfs.add(new MetricFamilySamples("cloudwatch_exporter_scrape_error", Type.GAUGE, "Non-zero if this scrape failed.", samples));
+
+      String buildVersion = "";
+      String releaseDate = "";
+      int errorFlag = 0;
+      try {
+          final Properties properties = new Properties();
+          properties.load(CloudWatchCollector.class.getClassLoader().getResourceAsStream(".properties"));
+          buildVersion = properties.getProperty("BuildVersion");
+          releaseDate = properties.getProperty("ReleaseDate");
```
Comment (matthiasr) [inline src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java:698]: The build info, by its nature, never changes. Could we do this work up-front and only set the metric once, or otherwise store the values to avoid unnecessary per-scrape work?

Comment (djloude) [reply src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java:713]: Yes I can move it up, and stop committing to the master before its ready to be looked at again! Sorry

Comment (matthiasr) [reply src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java:713]: All good, I was just passing through and couldn't keep from commenting 😬. Let me know when you want another round!

**Review by matthiasr — COMMENTED** [review]
This looks good, but there are other commits mixed into this PR now. Could you please rebase your fork, or even better create a feature branch and open a new PR from that?

File: README.md [hunk 7]
```diff
@@ -108,7 +110,23 @@ All metrics are exported as gauges.
...
+set to 1. If the metrics label values are "unknown" the build information scrap failed. 
+
+To display the metrics label values in Grafana:
+```
+1. Query for the cloudwatch_exporter_build_info
+2. Set the Legend to the desired label (ex. {{build_version}} or {{release_date}})
+3. Select Stat visualization
+4. Under stat styles select Name
+5. To remove the graph, select Graph Mode None
+```
```
Comment (matthiasr) [inline README.md:129]:
```suggestion

```

This changes every few Grafana versions, let's not duplicate Grafana documentation here.

**Comment by matthiasr** [comment]
Sorry for the delay, I was on vacation for a bit. I think the collector itself is good to go; I would like to keep the Grafana specifics out of the README, and you need to rebase this to only contain your commit(s) (best squash all your commits into one while you are at it). Once you do that I'll merge it @djloude!

**Comment by matthiasr** [comment]
Thank you! I've been a bit busy lately but will take a look as soon as I
can!

**Comment by matthiasr** [comment]
Replaced by #400
